src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs

```diff
@@ -209,8 +209,13 @@ internal static bool HasElementType(RuntimeType type)
             return outHandles;
         }
 
-        internal static object CreateInstanceForAnotherGenericParameter([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor | DynamicallyAccessedMemberTypes.NonPublicConstructors)] RuntimeType type, RuntimeType genericParameter)
+        internal static object CreateInstanceForAnotherGenericParameter(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] RuntimeType type,
+            RuntimeType genericParameter)
         {
+            Debug.Assert(type.GetConstructor(Type.EmptyTypes) != null,
+                $"CreateInstanceForAnotherGenericParameter requires {nameof(type)} to have a public parameterless constructor so it can be annotated for trimming without preserving private constructors.");
```

Comment:
I don't think the desire to put some annotation on the method is enough to make it an assumption. I assume you checked that the runtime will actually call the parameterless .ctor (since we're not passing in any .ctor args here) - I think that would be a much more solid assumption to make - which then leads to the better annotation.

Reply:
I'm not sure I understand your comment. I changed all callers of this method to ensure they have a public parameterless constructor.

This assert is to catch the cases where someone calls `CreateInstanceForAnotherGenericParameter` passing in a `Type` with an internal or private default constructor. That way they get an assert telling them this isn't supported.

Would it be clearer if I added

```suggestion
            Debug.Assert(type.GetConstructor(Type.EmptyTypes) is ConstructorInfo c && c.IsPublic,
                $"CreateInstanceForAnotherGenericParameter requires {nameof(type)} to have a public parameterless constructor so it can be annotated for trimming without preserving private constructors.");
```

?

Reply:
Got it - it's not about "parameterless" but about "public"
Makes sense.